File: src/floodFill.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) 
    {
        if (image[sr][sc] == color)
        {
            return image;
        }
        int row = image.size();
        int column = image[0].size();

        stack<pair<int, int>> stack;
        vector<vector<bool>> visited(row, vector<bool>(column, false));
        
        stack.push({ sr, sc });
        visited[sr][sc] = true;

        while (!stack.empty())
        {
            pair<int, int> cord = stack.top();
            stack.pop();
            int oldColor = image[cord.first][cord.second];
            image[cord.first][cord.second] = color;

            if (cord.first + 1 < row && !visited[cord.first + 1][cord.second] && image[cord.first + 1][cord.second] == oldColor)
            {
                stack.push({ cord.first + 1, cord.second });
                visited[cord.first + 1][cord.second] = true;
            }
            if (cord.first - 1 >= 0 && !visited[cord.first - 1][cord.second] && image[cord.first - 1][cord.second] == oldColor)
            {
                stack.push({ cord.first - 1, cord.second });
                visited[cord.first - 1][cord.second] = true;
            }
            if (cord.second + 1 < column && !visited[cord.first][cord.second + 1] && image[cord.first][cord.second + 1] == oldColor)
            {
                stack.push({ cord.first, cord.second + 1 });
                visited[cord.first][cord.second + 1] = true;
            }
            if (cord.second - 1 >= 0 && !visited[cord.first][cord.second - 1] && image[cord.first][cord.second - 1] == oldColor)
            {
                stack.push({ cord.first, cord.second - 1 });
                visited[cord.first][cord.second - 1] = true;
            }
         
        }
        return image;

    }
};

File: src/LeetCodePY/floodFill.py
from collections import deque
from typing import List

class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:

        if not image:
            return []

        if image[sr][sc] == color:
            return image

        oldColor = image[sr][sc]

        queue = deque()
        queue.append((sr, sc))
        
        while queue:
            x, y = queue.popleft()
            image[x][y] = color

            if x + 1 < len(image) and image[x + 1][y] == oldColor:
                queue.append((x+1, y))
            if x - 1 >= 0 and image[x - 1][y] == oldColor:
                queue.append((x - 1, y))
            if y + 1 < len(image[0]) and image[x][y + 1] == oldColor:
                queue.append((x, y+1))
            if y - 1 >= 0 and image[x][y - 1] == oldColor:
                queue.append((x, y-1))

        return image